### PR TITLE
Feature/functional test payload push

### PR DESCRIPTION
**Pull request by monkeydioude**

### CODE REVIEW

**Comment by monkeydioude** [comment]
WIP time is over @pyxide 
enjoy reviewing xD

File: build/send-payload/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+## send-payload image ##
+FROM alpine
+COPY ./testdata/send-payload /send-payload
+WORKDIR /src
+RUN apk add curl openssl jq
+ENTRYPOINT /send-payload $0 $1 $2
```
Comment (pyxide) [inline build/send-payload/Dockerfile]: Does it work with `/send-payload "$@"` ? It would allow any number of options.

File: josuke.go [hunk 2]
```diff
@@ -40,10 +41,9 @@ func parseLogLevel(value string) (LogLevel, bool) {
 // Josuke is the main object, that contains the HTTP server configuration
 // and the hook definitions.
 type Josuke struct {
-	LogLevelName string   `json:"logLevel"`
+	LogLevelName string `json:"logLevel"`
```
Comment (pyxide) [inline josuke.go]: Format nitpicking, not aligned with the below options.

File: script/docker-run.sh [hunk 3]
```diff
@@ -31,23 +33,32 @@ fi
 # Checking image already has a running container
 CONTAINER_ID=$(docker ps -qf ancestor="$BIN_IMAGE_NAME")
 if [ ! $CONTAINER_ID = "" ]; then
-    echo "[ERR ] Container with ID '$CONTAINER_ID' already running for image '$BIN_IMAGE_NAME'"
+    echo "[ERR ] container with ID '$CONTAINER_ID' already running for image '$BIN_IMAGE_NAME'"
     exit 1
 fi
 
-PORT=$(jq '.port' "$CONF_FILE")
+PORT=$(cat "$CONF_FILE" | docker run -i imega/jq '.port')
```
Comment (pyxide) [inline script/docker-run.sh:40]: You are hereby condemned to burn in hell as long as you will run the (all)mighty `jq` in Docker.

File: script/docker-run.sh [hunk 4]
```diff
@@ -2,12 +2,14 @@
 
 u_ok_jojo() {
     maxIt=4
-    sleepDuration=1
+    sleepDuration=3
+    sleep $sleepDuration
     for d in `seq 1 $maxIt`; do 
-        if [ ! $(docker ps -qf ancestor="$BIN_IMAGE_NAME") = "" ]; then
+        containerID=$(docker ps -qf ancestor="$BIN_IMAGE_NAME")
+        if [ ! $containerID = "" ] && [ $(docker inspect --format="{{.State.Health.Status}}" $containerID) = "healthy" ]; then
```
Comment (pyxide) [inline script/docker-run.sh]: I prefer `if [ -n "$containerId" ]`

File: script/docker-run.sh [hunk 5]
```diff
@@ -31,23 +33,32 @@ fi
...
+fi
+
+containerID=$(docker run --network="host" $RM_FLAG -d -v $(pwd):/src -e "CONF_FILE=$CONF_FILE" -e "PORT=$PORT" $BIN_IMAGE_NAME)
+echo "[INFO] container running with untruncated ID '$containerID'"
 
 # checking container status
 u_ok_jojo
 if [ $? = 1 ]; then
-    echo "[ERR ] Container did not start"
+    echo "[ERR ] container did not start properly (not running or unhealthy)"
```
Comment (pyxide) [inline script/docker-run.sh:59]: Condition should be `if [ ! $? = 0 ]`

File: josuke.go (hunk 2 above)
Comment (monkeydioude) [reply josuke.go]: This is how `gofmt` works, dunno if this is intended tho. The line without annotation below creates a separation in the indenting mechanism of the whole structure.
Will move the unannotated line for Adrian Monk' sake

File: script/docker-run.sh (hunk 3 above)
Comment (monkeydioude) [reply script/docker-run.sh:40]: :D
I agree `jq` is (somehow)mighty, that is the reason I use it. I wish it was as standard as `sed` or `awk` so I would not have to do this

File: build/send-payload/Dockerfile (hunk 1 above)
Comment (monkeydioude) [reply build/send-payload/Dockerfile]: I thought I tried and it didnt work, but I tried again and now, as King Crimson stand:

Comment (pyxide) [reply build/send-payload/Dockerfile]: Don’t be a schizoid man, it happens to everybody.

Comment (pyxide) [reply build/send-payload/Dockerfile]: I wanted to ask you to add quotes around `"$@"`, but when I take a look at https://github.com/imega-docker/jq/blob/master/Dockerfile, it reminds me that you don’t need to use `/send-payload $@`, simply use `/send-payload`, Docker will pass the arguments to the entrypoint script.

```Dockerfile
FROM alpine:3.10
ENTRYPOINT ["seq"]
```

```sh
# docker build -t seq .
# docker run -i seq 1
1
# docker run -i seq 1 2
1
2
# docker run -i seq 1 2 5
1
3
5
```

Comment (monkeydioude) [reply build/send-payload/Dockerfile]: interesting: it ~~just~~ works only when `ENTRYPOINT` is fed array like parameters. Would not work with `ENTRYPOINT seq` shell like params

**Review by oparent-sharecare — APPROVED** [review]
This is fantastic.

**Review by oparent-sharecare — APPROVED** [review]

**Review by pyxide — APPROVED** [review]
This is really fantastic!

**Pull request merged**